### ISSUE TITLE
DC2 Storage Discussion

### ISSUE THREAD

**Issue by salmanhabib** [LSSTDESC/DC2-production#52]
It would be good to have a discussion of the DC2 storage issue here. Initial info was provided by Simon Krughoff and later discussions with Jim Chiang and Mike Jarvis were very helpful (I am using a lot of information below provided by Mike).

**Issue** There is a question of why the storage requirement prefactor is so large: 15X where X is the raw file size. 

The explanation from Simon (paraphrased) is as follows: 

**Explanation** 2x: int --> double; another 2.5x --> adding mask plane and variance plane. Plus copies of each of the following data products: calexp (calibrated exposure), warp (warped calibrated exposures), coadd (coadded warped calibrated exposures). Calexps and warps are the same size, coadds are smaller than both. Roughly speaking, if one assumes that the coadds plus all the other data: e.g. output catalogs, configs, logs, is similar in size to the calexps and warps, this means another factor of 3. Thus one arrives at: 3*(2.5*(2raw_filesize)) = 15raw_filesize uncompressed. 

**Discussion** The data is intrinsically int32 and the initial factor of 2 comes from python's float64 (_this is incorrect, see Jim's comment below_) but there is no reason to store int32 as float64, so this factor of 2 can be immediately eliminated. The mask plane is integer and sparse so should be reducible to a much smaller size (lossless Rice compression). The variance plane does not require high precision and can be subject to lossy compression (reduction possibly by a factor of 4). Mike was not sure why we need to keep the calibrated exposure and was even more convinced that we do not need to keep the warped calibrated exposure. The coadds are needed.

Thus, it appears that the actual storage requirement is likely to be a multiplier of a "few" times the raw file size, where "few" is likely 2-3, and the raw file size is whatever it is in int32. This is a very big saving over what we are doing now.

Simon says that the DM stack has an abstraction layer that determines how data is stored. We could work with the DM folks to modify this or, alternatively, write our own data reduction code via a script that runs after each command line task (suggested by Jim, who believes that this should be straightforward).

**Comment by jchiang87** [comment]
A couple of clarifications:  from looking at the DC1 data in

/global/projecta/projectdirs/lsst/production/DC1/DM/DC1-imsim-dithered

the image data in the calexps, coadds, etc produced by the Stack are stored as float32 not float64, so that explanation for the initial factor of two wrt the raw data is incorrect.   BTW, the mask plane data are int16 and the variance plane is float32, (assuming that the hdu ordering is 1->image, 2->mask, 3->variance....frustratingly, the Stack doesn't name the extensions in the output FITS files), so the next factor of 2.5 is manifestly correct. 

I believe the initial factor of two does not arise from int32 -> float64, but rather from the fact that the eimages start out compressed (via gzip) and so are ~1.7-2 smaller than their uncompressed sizes.

**Comment by salmanhabib** [comment]
@jchiang87 Ah, ok, good catch on the initial factor of two. I had not assumed any compression, but the typical lossless compression factor should be as you say.

**Comment by cwwalter** [comment]
> Mike was not sure why we need to keep the calibrated exposure and was even more convinced that we do not need to keep the warped calibrated exposure. The coadds are needed.

We did use the calibrated exposures in our DC1 work for several things.   We want to see how things are working on an exposure by exposure basis, not just the co-adds and we anticipate using them again.  That is also where you grab the calibration info from etc.  For algorithm development I think it important to have them.

For the warped exposures: I agree mostly these aren't used.  We think they might be useful if people wanted to do work comparing what is in the co-add with the constituent exposures.  But, they can be recreated if necessary from the calexps.  My suggestion was that we only keep one patch on disk so people can play with them and then if they need to others either recreate them on the fly with the calexps or (maybe better) store them compressed patch-by-patch on tape so they could be called up as needed.

One thing that is missing is that there was a directory (I think it was called "icExp"?) that was basically just temp files.  It took something like a third of the space.  We decided to just delete that.

**Comment by cwwalter** [comment]
FYI I found the conversation in slack where we discussed the directories.  I did a diskuse check of the directories that looked like this:

```
cori08:DC1-imsim-dithered % du -sh * | sort -hr
28T     icExp
28T     calexp
21T     deepCoadd
2.0T    src
195G    icSrc
188G    deepCoadd-results
12G     processEimage_metadata
2.6G    srcMatch
838M    eimage
243M    ref_cats_orig
125M    deep_makeCoaddTempExp_metadata
17M     registry.sqlite3
9.9M    background_values
4.9M    deep_assembleCoadd_metadata
620K    config
408K    schema
4.0K    _mapper
0       ref_cats
```

Go here to see more:

https://lsstc.slack.com/archives/C2M6H2LAX/p1501867673418744

**Comment by salmanhabib** [comment]
I asked Rachel to point the relevant science WG folks at this discussion to see who wants what kept (and at what level of accuracy). Meantime we should think about the data reduction layer implementation issue.

**Comment by cwwalter** [comment]
> Meantime we should think about the data reduction layer implementation issue.


There is now the ability to do compression on the fly in the butler, so some work on this is already happening but I'm not sure of all of the details.  We can find out more from the DM people.

**Comment by rmandelb** [comment]
@salmanhabib - I am more than happy to point people at this thread and make sure we get some feedback (I would target people who I know have some particularly demanding analyses in mind) but is there some place that lists what exactly are the things that we should consider not keeping?  Reading through the thread I see references to sizes of the following:

- calexp (calibrated exposure), 
- warp (warped calibrated exposures), 
- coadd (coadded warped calibrated exposures) 
- mask plane
- variance plane

We'll keep the coadd, so we don't need to ask people about that, but at least the need for calexps and warped calexps has been called into question as I understand it.  Is there anything else that we may consider discarding or storing at lesser precision?  I just want to have as clear a question as possible for the people who I draw into this discussion.

**Comment by jchiang87** [comment]
Groups like SN and SL will have some interest in keeping the calibrated exposures around since they may want to do their own flux extractions to get better light curves than what the standard DIA processing provides.  We can reduce the storage of the calibrated exposures by just keeping smaller postage stamps of the fields for the time variable objects, but this would require software development on our part, and I'm not sure the Stack itself can work out-of-the box on such postage stamps (assuming those WG would want to use the Stack for their additional image analyses).   I'm not sure it's a good idea to discard the mask and variance planes independently of the image data they are associated with, but we can certainly use a different (i.e., lossy) compression scheme for the variance planes.

**Comment by salmanhabib** [comment]
@rmandelb We did not want to drop the mask and variance planes, just to compress them (they ought to be highly compressible). No one has indicated a requirement for keeping warp so far.

**Comment by cwwalter** [comment]
I suggested at least keeping a subset (say a patch) of them and the ability to bring others back from tape in patches.  It could be useful (for example) for people comparing algorithms where you are acting on the exposures individually with the co-add because otherwise they don't line up.

**Comment by salmanhabib** [comment]
@cwwalter Is there a specific reason to keep the warp info? Which SWG wants this and why? We can always do special smaller runs later on for special tasks.

**Comment by rmandelb** [comment]
Hi Salman -

> We did not want to drop the mask and variance planes, just to compress them

Good, this is consistent with my thinking.  Now I'll ping a few people who I think might be able to say exactly what they need for already-planned analysis.

**Comment by rmandelb** [comment]
@esheldon @rmjarvis @joezuntz -

For the DC2 simulations, there is some question of what intermediate data products created during the DM processing need to be saved once the DM processing is complete in order to enable the planned DC2 analyses.  We assume we will keep the coadd, and some compressed version of the mask and variance plane.  What about the calexps (calibrated exposure) and warps (warped calibrated exposures)?  Do you envision any use for them for the WL group work on testing the shear pipeline?

I was wondering if you intend, for example, to do tests of per-object coadds or any of the stuff that Erin has been testing on HSC, and whether that leads to a need for calexps or warps?

**Comment by rmandelb** [comment]
@rbiswas4 @DominiqueFouchez -

For the DC2 simulations, there is some question of what intermediate data products created during the DM processing need to be saved once the DM processing is complete in order to enable the planned DC2 analyses. We assume we will keep the coadd, and some compressed version of the mask and variance plane. What about the calexps (calibrated exposure) and warps (warped calibrated exposures)?  What do you think you will need kept from the uDDF simulations for the SN group work on image-based classification?

**Comment by katrinheitmann** [comment]
@rmandelb @salmanhabib All, great to see that all this information is collected! When you have arrived at a good point to summarize this, there is Section 10 in the planning document where this could all go. I think it would be very helpful to put it there and we can point anybody interested there to find a summary. Thanks!

**Comment by rmandelb** [comment]
@katrinheitmann - sounds good.  Since I've deliberately pinged people who are planning some of the more data-intensive analyses but haven't pinged everybody (which I think would lead to chaos at this stage), it is definitely a good idea to write up what we tentatively agree on based on the needs for these applications, and then ask for comments from the other analysis WGs.

One more thing: it's possible our answers will differ for uDDF vs. WFD.  I assume this is not a problem, but please let me know if it is.

**Comment by DominiqueFouchez** [comment]
@rmandelb, @rbiswas4  For SN, in order to redo a photometry we need all the info of the calibrated exposure (calex, variance, mask). We need also all the difference images. A priori no warps if we can redo it (to undertand any astrometric problem). This is true also for uDDF.

**Comment by SimonKrughoff** [comment]
Just to talk to some of the conversation here.
* I believe we can turn off the saving of the `icExp` products completely, so it wouldn't even require an afterburner to remove them.
* I'm not sure what processing people want to do on postage stamps.  It's possible, but many algorithms require some knowledge of the larger context in which a single objects reside.
* A possible reason to keep the warps is that it makes it trivial to make a time series movie for a given area of sky.  I can't think of a science motivated reason to keep them.
* Compression is an option, but I think we should probably use standard tools to do that (gzip, fpack) rather than pulling the images apart and compressing the mask separate from the variance.

**Comment by cwwalter** [comment]
> Is there a specific reason to keep the warp info? Which SWG wants this and why? We can always do special smaller runs later on for special tasks.

Well.. for algorithm work we need the calexps.  The question is whether we also need the warps. I'm not positive but when I tried working with DC1 and wanted to see the exposures that went into a co-added patch with the butler what you get is the list of warps (because that is what is used to make it).

So, Simon and I and others talked about this before.  It isn't prohibitive to just remake them if you want (but it is also not free). The suggestion I had was that we keep one patch on disk (which takes up little space) which will allow people to test their algorithms on a smaller scale and then if they need more we tar the other patches to tape to people could bring them up to the scratch space if needed.

If the tape is also an issue we could setup something for them to regenerate them if they are needed later.

**Comment by rmjarvis** [comment]
> Compression is an option, but I think we should probably use standard tools to do that (gzip, fpack) rather than pulling the images apart and compressing the mask separate from the variance.

I don't really understand this comment.  Yes, we should use fpack, but that already does separate compression for each HDU in the fits file.  So why would we need to pull apart the image to do this.

Also, I'm frankly surprised that compression isn't just being done as a matter of course by LSSTDM.  It just seems gratuitously wasteful to not at least to lossless compression.  And lossy Rice compression should be completely fine with around 4 or 5 noise bits (q=4 or 8), which would be far more compressive and add an unnoticeable amount of extra noise.

**Comment by SimonKrughoff** [comment]
> I don't really understand this comment. Yes, we should use fpack, but that already does separate compression for each HDU in the fits file. So why would we need to pull apart the image to do this.

I just couldn't tell by quickly scanning the thread what the proposal was.

> Also, I'm frankly surprised that compression isn't just being done as a matter of course by LSSTDM. It just seems gratuitously wasteful to not at least to lossless compression.

Lossless compression is now on by default, so this compression discussion may be moot.  See: [this RFC](https://jira.lsstcorp.org/browse/RFC-378) and [this implementation ticket](https://jira.lsstcorp.org/browse/DM-12599) for discussion.

**Comment by rmjarvis** [comment]
Probably not moot.  But it's at least improperly informed.  

We should look at what the actual file sizes are with the current default compression that you guys are now using rather than what we saw for DC1 (where I guess there was not compression).  Then we can evaluate whether moving to lossy compression for some files is warranted.

**Comment by cwwalter** [comment]
> (where I guess there was not compression). Then we can evaluate whether moving to lossy compression for some files is warranted.

Correct.. no compression for DC1 (it wasn't an option then). After the fact the files were compressed so that they could be put to tape (I believe with bzip).  Mustafa could tell you what the achieved factor was if it was relevant.

**Comment by SimonKrughoff** [comment]
> Probably not moot. But it's at least improperly informed.

Fair enough.

I will also say that Robert Gruendl is currently leading a working group looking at lossy compression in the context of LSST data storage.  I'm not sure where that stands, but can ask around.

**Comment by rbiswas4** [comment]
@rmandelb @DominiqueFouchez @salmanhabib @reneehlozek 


> For the DC2 simulations, there is some question of what intermediate data products created during the DM processing need to be saved once the DM processing is complete in order to enable the planned DC2 analyses. We assume we will keep the coadd, and some compressed version of the mask and variance plane. What about the calexps (calibrated exposure) and warps (warped calibrated exposures)? What do you think you will need kept from the uDDF simulations for the SN group work on image-based classification?

Unfortunately, I am not sure of the requirements we will have. But here is a first stab at this. Hopefully someone else can comment on this and we can start making progress. 

First, I assume you are exploring the possibility of dropping these auxiliary products after running through the DiaObject creation pipeline (ie. Running Difference imaging, source detection and photometry on the difference, and then doing forced photometry on differences at least within a time range). If the running of any of these steps is on us, then of course, we will need whatever  auxiliary products the DM stack needs to run the remaining steps.  So, if there are DM steps in the transient pipeline that will not be run before this is thrown away, we need to find out from DM what is allowed.

One of the major things we would like to run separately is an off-the-shelf version of artifact-source classifier. In its current state, it only runs on individual difference image cutouts (so the warps are not necessary). A clear improvement would be in running this on difference image cutouts from multiple visits + the registered template / science image. It would be good to check if this is possible without the warps or if the warps can be generated quickly enough from the calexps that it does not matter either way. Second, for the uDDF one of the extra things we plan to do is perform coadds of the visits before doing the diffims. Again I think not having the warps might prevent this, but again I would like to verify this with others.

In terms of lossy compression, I don't have an estimate of the requirement right now, but I think we know how to put a requirement. We will be working hard to reach a calibration requirement of a few 10^{-3}. If we want to study our precision, the extra stuff due to compression should be less that this, say a factor of 10+. If there is a real problem, it may be ok to relax this  bit during DC2 only, as the best calibration modules are not available either.

**Comment by salmanhabib** [comment]
One general maxim to keep in mind is that "compute is free" but storage is not. So it's fine to be able to recompute things later on provided the (subset of) data needed to do this is easily accessible.

**Comment by rmjarvis** [comment]
> We assume we will keep the coadd, and some compressed version of the mask and variance plane. What about the calexps (calibrated exposure) and warps (warped calibrated exposures)? Do you envision any use for them for the WL group work on testing the shear pipeline?

To get back to this (on prompting by Rachel...), my best guess from the WL perspective is to keep the original (int, not DM-expanded into float) images, the coadds, and I think the calexps.  Also (obviously) all the metadata like WCS solutions and zero-point values.

I don't really know the LSSTDM lingo, but I'm assuming calexp is bias, dark, and flat-corrected, which is what we want.  Probably also want it sky subtracted, since that needs to be done carefully when the flats are mostly astrometric patterns, not QE patterns, so just dividing by the flat is wrong.  I'm sure Robert and co already know all this.  I'm just clarifying that we want the end result of all that saved.  Probably nothing prior to that.

We also don't need anything subsequent to that (the warps) that are used in making the coadds.

Finally, I would recommend using Rice compression on the calexp images with q=4 or 8 (i.e. keep 4 or 5 bits of noise).

Again, that's from a WL perspective, which is probably also applicable to most other non-time-domain science.  YMMV for SN, TMO, etc.

**Comment by boutigny** [comment]
@cwwalter icExp are related to the CharacterizeImageTask and their output can be turned off. In fact I think that the default is to not write them out. Otherwise, my experience is that calexps are useful and warps are useless. 
For instance, regarding calexps, @esheldon should confirm, but I am almost sure that they are necessary to produce meds files. They are also very useful to debug and understand what happens at the individual image level.
As for compression I had a very bad experience when I tried to turn on supposedly "lossless" fits compression on calexp and tried to make coadd from the compressed files (I don't remember the details). We can probably make it work but we should be very cautious.

**Comment by esheldon** [comment]
We do want the calexp for making the MEDS files used for various codes such as the metacalibration shear code, BFD shear code, and ngmix/MOF photometry and deblending codes.  However, since we now actually plan to make "per object" coadds, the input could be the images used to make the standard coadd (called the warps?) since we would have to create those in any case.

If the above are easy and fast to generate "on the fly" we might not need to store them, but I don't know if that is the case.

**Comment by boutigny** [comment]
Interesting ! So it means that it could be useful to keep the warps. They are not very cpu intensive to create but the I/O rate is pretty high. I just had a look to the HSC processing at IN2P3, 15 patches (4000x4000) are processed in 3.5 hours (wall clock) with a ratio CPU/Wall ~4%.

**Comment by salmanhabib** [comment]
@boutigny I am surprised you ran into problems with lossless compression. Is this something to do with fits, in particular?

@esheldon and @rmjarvis can you two please come to a consensus on warps yes/no? Thanks!

**Comment by esheldon** [comment]
We will I think want to use warps, so it is a technical question if we can generate them fast enough.  I think from Rachel's post it seems they are quite slow to generate, so we will I think want to store them.

**Comment by esheldon** [comment]
actually I think she must not have posted that here:  I think she checked some log files and saw that during generation the process was i/o bound and overall very slow.

**Comment by rmandelb** [comment]
Phew, I was wondering if I was posting in here in my sleep or if perhaps I have an imposter.  (I actually didn't say that at all anywhere...)  Then I realized you may be referring to Dominique's post above...?

**Comment by esheldon** [comment]
Yes, sorry, that was from Dominique

**Comment by boutigny** [comment]
This is from the actual Dominique :-) The time to produce the warps is not negligible but not that long. Then it depends if you expect to have to reproduce the warps many time in which case it is certainly better to keep them on disk. If you need them in a few occasions only, then I would recommend to get rid of them and to reproduce them when needed. 
But reproducing the warps supposes that you have kept the calexps

**Comment by rmjarvis** [comment]
Given that the warps are relatively easy to make from calexps, but not vice versa, I suggest we keep the calexps and not the warps after the coadds are made.

If we can do both kinds of coadds at the same time, then we probably never need to remake the warps.  But even if the per-object coadds are done later than the detection coadds, making the warps should be a relatively small portion of the time needed to make those, so it shouldn't be a large burden to not have them spinning already.

**Comment by rmandelb** [comment]
My reading of the comments above is that the SN image-based analysis and a limited subset of the WL analysis may need the warps, but (a) this isn't guaranteed (at least for SN), (b) there is a decent amount of work they can do without the warps, (c) we can regenerate them from calexps (as Dominique B said the process is a bit i/o-heavy but not super expensive).  In contrast, the calexps are needed for more of their work at an earlier stage and I would be inclined based on that to keep them.

Chris had mentioned the possibility of keeping the warps for a small patch.  I think this might be a good idea; people can use that to set up their pipelines and do some limited tests, before deciding that it's worthwhile to go the route of regenerating the warps for whatever science application for a far larger area.

So this means we want to keep:
- original images (Mike specified we can keep the int versions, not DM-expanded to floats, and I don't see why this wouldn't work for everybody)
- coadds
- calexps
- metadata like WCS solutions and zero-point values
- variance, mask
- There was a request from Dominique F for the difference images to be saved.

And then there is presumably a good amount of savings due to compression, with a scheme that is TBD.

Are there any objections to saying this has converged?  I didn't ping every single analysis WG, but the ones I did ping are likely to have the most intensive hands-on work that is closest to the pixels (the others are likely to stay farther from the pixels or be doing similar work as one of the ones I did ping).  I am happy to write this into section 10 of the writeup if helpful.

**Comment by katrinheitmann** [comment]
It would be absolutely helpful if you can capture this in Section 10 -- thanks a lot!

**Comment by rmandelb** [comment]
I see Mike has thumbs-uped my message.  I would like to give one more day for final comments and then I will indeed put it in section 10.

**Comment by rmandelb** [comment]
For the record, I put this in section 10 a few days ago (didn't want to make a lot of github traffic at the time so didn't comment at that point).  

@salmanhabib - I did not close this issue because I thought the question of what compression method will be used has not been resolved.

**Comment by cwwalter** [comment]
> Chris had mentioned the possibility of keeping the warps for a small patch. I think this might be a good idea; people can use that to set up their pipelines and do some limited tests, before deciding that it's worthwhile to go the route of regenerating the warps for whatever science application for a far larger area.

To be a bit more explicit: I think keeping the warps from a patch worth of calexps on disk would be good since then you would have all of the inputs that went into a co-add for testing and studies.

I also suggested we might want to tar up the other co-adds by patch and put them to tape if people needed them (rather then regenerating) but I don't know if:

- Using the tape is a cost driver
- Whether people will really need them.  If we don't know we also might want to wait and then if we do need to regenerate them store them to tape at that time.

**Comment by salmanhabib** [comment]
@rmandelb Let's discuss the compression alternatives next week, may be worth a separate threads for test results, etc.
@cwwalter Tape is not a cost driver for the amount of data we are talking about.

**Comment by salmanhabib** [comment]
HI Everyone, I would like to have a conclusion here so that we can plan what we need to do for Run 1.1.
The summary should be captured in a short DESC Note.

**Comment by SimonKrughoff** [comment]
@salmanhabib I think we need to action someone to do the note.  Do you have a suggestion?  I can help, but probably shouldn't be the main author.

**Comment by katrinheitmann** [comment]
This issue has been quiet for more than 2 years. We had lots more discussions about storage in the meantime. We might need to open up a new issue once we have all data products from DC2 that we want to store longer term and also to make plans for actual data (@heather999 and @JoanneBogart ). But I think this issue can be closed for now. Feel free to reopen if you think otherwise. Thanks!
